connectionmanager/connectionmanager.go

```diff
@@ -59,72 +59,67 @@ type ProtocolHandler struct {
 
 var handlersByTypeMutex = &sync.Mutex{}
 var handlersByType = map[ws.ProtoType]*ProtocolHandler{}
-var reconnectIntervalSeconds = 5
 var DefaultPingWait = time.Minute
 
 func GetWriteTimeout() time.Duration {
 	return writeWait
 }
 
-func SetReconnectIntervalSeconds(i int) {
-	reconnectIntervalSeconds = i
+var a *expBackoff = &expBackoff{
+	maxInterval:  60 * time.Minute,
+	attempts:     0,
+	exceededMax:  false,
+	maxBackoff:   120 * time.Minute,
+	smallestUnit: time.Minute,
 }
 
 func connect(
 	proto ws.ProtoType,
 	serverUrl, connectUrl, token string,
-	retries uint,
 	ctx context.Context,
 ) error {
 	parsedUrl, err := url.Parse(serverUrl)
 	if err != nil {
 		return err
 	}
-
 	scheme := getWebSocketScheme(parsedUrl.Scheme)
 	u := url.URL{Scheme: scheme, Host: parsedUrl.Host, Path: connectUrl}
 
 	cancelReconnectChan[proto] = make(chan bool)
 	var c *connection.Connection
-	var i uint = 0
 	defer func() {
 		setReconnecting(proto, false)
 	}()
+
 	setReconnecting(proto, true)
-	for IsReconnecting(proto) {
-		i++
-		c, err = connection.NewConnection(u, token, writeWait, maxMessageSize, DefaultPingWait)
-		if err != nil || c == nil {
-			if retries == 0 || i < retries {
-				if err == nil {
-					err = errors.New(
-						"unknown error: connection was nil but no error provided by" +
-							" connection.NewConnection",
-					)
-				}
-				log.Errorf("connection manager failed to connect to %s%s: %s; "+
-					"reconnecting in %ds (try %d/%d); len(token)=%d", serverUrl, connectUrl,
-					err.Error(), reconnectIntervalSeconds, i, retries, len(token))
-				select {
-				case cancelFlag := <-cancelReconnectChan[proto]:
-					log.Tracef("connectionmanager connect got cancelFlag=%+v", cancelFlag)
-					if cancelFlag {
-						return nil
-					}
-				case <-ctx.Done():
-					return nil
-				case <-time.After(time.Second * time.Duration(reconnectIntervalSeconds)):
-					break
-				}
-				continue
-			} else if i >= retries {
-				return ErrConnectionRetriesExhausted
+	c, err = connection.NewConnection(u, token, writeWait, maxMessageSize, DefaultPingWait)
+	if err != nil || c == nil {
+		if err == nil {
+			err = errors.New(
+				"unknown error: connection was nil but no error provided by" +
+					" connection.NewConnection",
+			)
+		}
+		// Calculates backoff time before next connection attempt
+		intvl := a.GetExponentialBackoffTime()
+		a.attempts++
+		log.Errorf("connection manager failed to connect to %s%s: %s; "+
+			"reconnecting in %s; len(token)=%d", serverUrl, connectUrl,
+			err.Error(), intvl, len(token))
+		select {
+		case cancelFlag := <-cancelReconnectChan[proto]:
+			log.Tracef("connectionmanager connect got cancelFlag=%+v", cancelFlag)
+			if cancelFlag {
+				return nil
 			}
+		case <-ctx.Done():
+			return nil
+		// Waits for the interval calculated by GetExponentialBackoffTime
+		case <-time.After(intvl):
 			return err
```

Comment:
Waiting to return the already available error is a waste of time. The waiting should happen before the call to establish the connection.